### ISSUE TITLE
Summary Counts Mixing

### ISSUE THREAD

**Issue by bradleygore** [neo4j/neo4j-go-driver#178]
I believe there is some sort of shared memory going on with the result summaries - the outcomes of an execution in a prior session seems to affect the numbers in the summary of a subsequent execution in a new session.

- Neo4j version: Enterprise 4.1.3
- Neo4j Mode: Single instance
- Driver version: `github.com/neo4j/neo4j-go-driver/v4 v4.2.0`
- Operating system: OSX
- Expected behavior: Summary counts for a query should not change based on prior queries executed in isolated session
- Actual behavior: Summary counts seem to be affected across executions even when in isolated sessions

Here is a minimal test file (using Ginkgo test framework and Gomega assertion lib, but should work w/ `go test` also), and below it are instructions for how to reliably reproduce the issue.

```go
package repos_test

import (
	"github.com/neo4j/neo4j-go-driver/v4/neo4j"
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/ginkgo/extensions/table"
	. "github.com/onsi/gomega"

	// here are exact versions of these items based on my go.mod:
	//github.com/neo4j/neo4j-go-driver/v4 v4.2.0
	//github.com/onsi/ginkgo v1.14.2
	//github.com/onsi/gomega v1.10.3

)

var _ = Describe("neo4j general", func() {
	BeforeEach(func() {
		sesh := driver.NewSession(neo4j.SessionConfig{AccessMode: neo4j.AccessModeWrite})
		defer sesh.Close()
		res, err := sesh.Run("OPTIONAL MATCH (f:FOO) DETACH DELETE f", nil)
		Expect(err).To(BeNil())
		Expect(res).NotTo(BeNil())
	})

	DescribeTable(
		"counters",
		func(cypher string, nodesCreated, nodesDeleted, relsCreated int) {
			var (
				sesh neo4j.Session = driver.NewSession(neo4j.SessionConfig{
					AccessMode: neo4j.AccessModeWrite,
				})
				result neo4j.Result
				err    error
			)
			defer func() {
				Expect(sesh.Close()).To(Succeed(), "sesh close failed")
			}()

			result, err = sesh.Run(cypher, nil)
			Expect(err).To(BeNil(), "got error when executing cypher")

			summary, err := result.Consume()
			Expect(err).To(BeNil(), "could not consume result to get summary")

			stmt := summary.Statement().Text()
			counters := summary.Counters()
			Expect(nodesCreated).To(Equal(counters.NodesCreated()), "nodesCreated not a match: %s", stmt)
			Expect(nodesDeleted).To(Equal(counters.NodesDeleted()), "nodesDeleted not a match: %s", stmt)
			Expect(relsCreated).To(Equal(counters.RelationshipsCreated()), "relsCreated not a match: %s", stmt)
		},
		Entry("should report one node created", "CREATE(f:FOO {id:123}) RETURN f", 1, 0, 0),
		Entry(
			"should report two nodes created and one relationship created",
			`CREATE(f:FOO {id:223})
			CREATE(f2:FOO {id:323})
			CREATE (f)-[:LINKS_TO]->(f2)`,
			2, 0, 1,
		),
		Entry(
			"should report zero nodes created/deleted and zero relationships created when nodes not found",
			`MATCH(f:FOO {id: 1122})
			MATCH(f2:FOO {id: 2211})
			CREATE (f)-[:LINKS_TO]->(f2)`,
			0, 0, 0,
		),
	)
})

```

When I run the above test file as-is, the first two `Entry`s pass successfully and the third one fails. Here is the output:

```
------------------------------
neo4j general counters 
  should report one node created
  /Users/bgore/git/authz/internal/repos/graph_test.go:63
•
------------------------------
neo4j general counters 
  should report two nodes created and one relationship created
  /Users/bgore/git/authz/internal/repos/graph_test.go:64
•
------------------------------
neo4j general counters 
  should report zero nodes created/deleted and zero relationships created when nodes not found
  /Users/bgore/git/authz/internal/repos/graph_test.go:71

• Failure [0.023 seconds]
neo4j general
/Users/bgore/git/authz/internal/repos/graph_test.go:10
  counters
  /Users/bgore/git/authz/internal/repos/graph_test.go:20
    should report zero nodes created/deleted and zero relationships created when nodes not found [It]
    /Users/bgore/git/authz/internal/repos/graph_test.go:71

    nodesDeleted not a match: MATCH(f:FOO {id: 1122})
                        MATCH(f2:FOO {id: 2211})
                        CREATE (f)-[:LINKS_TO]->(f2)
    Expected
        <int>: 0
    to equal
        <int>: 2
```

Notice that it thinks that there were 2 deleted nodes reported by the driver in this test - even though the query in question does not delete nodes. However, the `BeforeEach` does delete nodes. So I started exploring that a bit and I found that if I comment out the line `defer sesh.Close()` from the `BeforeEach`, the test then passes every time.

These tests create their own write session and run in isolation - e.g. starting with a blank slate and running sequentially but not depending on each other in any way (order could be randomized without issue), however it seems the write session summary counts from the `BeforeEach` are somehow making it into the test scenarios.

If I'm just structuring these incorrectly and the issue is on my end, please advise. However, I executed this same test file (modified slightly due to `NewSession` signature changing slightly) against the old version of the go driver (`github.com/neo4j/neo4j-go-driver v1.8.3`) and it succeeds every time without having to comment out the `defer sesh.Close()` in the `BeforeEach`.

**Comment by 2hdddg** [comment]
Hi @bradleygore thanks for your detailed bug report. You are right, the counters wasn't reset properly. A fix is on 4.2 branch and a patch release will be out soon.

**Comment by bradleygore** [comment]
Thanks for the quick resolution @2hdddg !
